yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -207,6 +211,17 @@ private[yarn] class YarnAllocator(
   }
 
   /**
+   * Gets the executor loss reason for a disconnected executor.
+   * Note that this method is expected to be called exactly once per executor ID.
+   */
+  def getExecutorLossReason(executorId: String): ExecutorLossReason = synchronized {
+    allocateResources()
+    // Expect to be asked for a loss reason once and exactly once.
+    assert(completedExecutorExitReasons.contains(executorId))
```

Comment:
Your assumption probably holds for the preemption case, since it's YARN killing the container. But I can imagine that if the container exits by itself, it might be possible for the disconnect to reach the driver endpoint and the `GetExecutorLossReason` message to reach the AM before the NM has had a chance to process the container exit and communicate that to the RM.